### PR TITLE
SSO with Keycloak

### PR DESCRIPTION
**Pull request by Loquacity**
https://github.com/SUSE/spacewalk/issues/11104

### CODE REVIEW

File: modules/administration/pages/auth-methods-pam.adoc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+----
+. Enforce the use of the service file by adding this line to [path]``/etc/rhn/rhn.conf``:
++
+----
+pam_auth_service = susemanager
+----
++
+In this example, the PAM service file is called [systemitem]``susemanager``.
+. Restart the {productname} services after a configuration change.
+. In the {productname} {webui}, navigate to [menuitem]``Create User`` and enable a new or existing user to authenticate with PAM.
```
Comment (keichwa) [inline modules/administration/pages/auth-methods-pam.adoc:31]:
```suggestion
. In the {productname} {webui}, navigate to menu:Users[Create User] and enable a new or existing user to authenticate with PAM.
```
"Create User" is probably wrong if you want to edit an existing user.  This probably must be done with the user's details page.  See reference:pages:users/user-details.adoc.  Hope I got it right.

File: modules/administration/pages/auth-methods-pam.adoc [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+Use the password change mechanism of your network service to change the password for these users.
+====
+
+
+To configure system-wide authentication you can use YaST.
+You will need to install the [package]``yast2-ldap-client`` and [package]``yast2-kerberos-client`` packages.
+
+
+For more information about configuring PAM, the SUSE Linux Enterprise Server Security Guide contains a generic example that will also work for other network-based authentication methods.
+It also describes how to configure an Active Directory Service.
```
Comment (keichwa) [inline modules/administration/pages/auth-methods-pam.adoc:51]: do we actually want to capitalize it?

File: modules/administration/pages/auth-methods-sso.adoc [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+
+
+.Procedure: Adding Endpoints as Clients
+. In the Keycloak {webui}, create a new client using these details:
++
+* In the ``Client ID`` field, enter the endpoint specified in the server configuration file as ``java.sso.onelogin.saml2.idp.entityid``.
+For example, ``https://<FQDN>/rhn/manager/sso/metadata``.
+* In the ``Client Protocol`` field, select ``SAML``.
+* Toggle the ``Includew AuthnStatement`` switch to ``On``.
```
Comment (keichwa) [inline modules/administration/pages/auth-methods-sso.adoc:177]:
```suggestion
* Toggle the ``Include AuthnStatement`` switch to ``On``.
```
Above it is spelled `AuthNStatement`.

File: modules/administration/pages/auth-methods-sso.adoc [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+
+
+When you have added the endpoints as clients, you can configure the client scope, and map the users between Keycloak and {productname}.
+
+
+
+.Procedure: Configuring Client Scope and Mappers
+. In the Keycloak {webui}, navigate to the menu:Clients[Client Scopes] tab and assign ``role_list`` as the default client scope.
+. Navigate to the menu:Clients[Mappers] tab and add a user attribute ``Uid`` mapper, using the default values.
```
Comment (keichwa) [inline modules/administration/pages/auth-methods-sso.adoc:197]: Is Uid correctly capitalized?  Above,we have many occurrencies of uid.

File: modules/administration/pages/auth-methods-sso.adoc [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+This SAML attribute is expected by {productname}.
+. Navigate to the menu:Users[Admin] section and create an administrative user.
+This user does not need to match the {productname} administrative user.
+. Navigate to the menu:Users[Role Mappings] tab, add a ``uid`` attribute with a value that matches the username of the {productname} administrative user.
+. Navigate to the menu:Users[Credentials] tab, and set the same password as used by the {productname} administrative user.
```
Comment (keichwa) [inline modules/administration/pages/auth-methods-sso.adoc:202]: strange. above we say that this admin user is different from the productname admin user.  Why using the same password?

File: modules/administration/pages/auth-methods-sso.adoc [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+. Save your changes.
+
+
+
+When you have completed configuration, you can test that the installation is working as expected.
+Restart the {productname} Server to pick up your changes, and navigate to the to {productname} {webui}.
```
Comment (keichwa) [inline modules/administration/pages/auth-methods-sso.adoc:208]:
```suggestion
Restart the {productname} Server to pick up your changes, and navigate to the {productname} {webui}.
```

File: modules/administration/pages/auth-methods-pam.adoc (hunk 2 above)
Comment (Loquacity) [reply modules/administration/pages/auth-methods-pam.adoc:51]: no, I don't think so.

File: modules/administration/pages/auth-methods-sso.adoc [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+The endpoint for ACS in {productname} is: https://example.com/rhn/manager/sso/acs
+* Single logout service (or SLS): an endpoint to initiate a logout request from the IdP.
+The endpoint for SLS in {productname} is: https://example.com/rhn/manager/sso/sls
+* Metadata: an endpoint to retrieve {productname} metadata for SAML.
+The endpoint for metadata in {productname} is: https://example.com/rhn/manager/sso/metadata
+
+
+[IMPORTANT]
+====
+Your IdP must have a SAML:Attribute containing the username of the IdP user domain, called `uid`.
```
[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:46]: agreed.

File: modules/administration/pages/auth-methods-sso.adoc [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+When you have prepared the {productname} Server, you can install Keycloak.
+You can install Keycloak directly on your machine, or run it in a container.
+In this example, we run Keycloak in a Docker container.
+For more information about installing Keycloak, see the Keycloak documentation at https://www.keycloak.org/getting-started/getting-started-docker.
+
+
+
+.Procedure: Setting Up the Identity Service Provider
+. Install KeyCloak in a Docker container, according to the Keycloak documentation.
```
[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:149]: Oof, I keep doing that by some weird force of habit (might be a leftover muscle memory from OpenStack? Not sure)

File: modules/administration/pages/auth-methods-sso.adoc (hunk 3 above)
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:177]: it should be a lowercase n

File: modules/administration/pages/auth-methods-sso.adoc (hunk 4 above)
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:197]: it's all over the shop in the source content. Looks like the keycloak UI uses `Uid` mostly.

File: modules/administration/pages/auth-methods-sso.adoc (hunk 5 above)
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:202]: That's probably a @srbarrios question. My understanding was that you create a user (that can be basically anything), and then give it the appropriate parameters so it authenticates correctly to SUMA.

File: modules/administration/pages/auth-methods-sso.adoc [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+This example is provided for illustrative purposes only.
+{suse} does not recommend or support third-party identity service providers, and is not affiliated with Keycloak.
+For Keycloak support, see https://www.keycloak.org/.
+====
+
+
+
+.Procedure: Setting Up the {productname} Server
+. On the {productname} Server, open the configuration file at ``/etc/rhn/rhn.conf`` and locate and edit these parameters.
+Replace ``FQDN`` with the fully-qualified domain name of your {productname} installation:
```
[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:117]: ?

File: modules/administration/pages/auth-methods-sso.adoc (hunk 4 above)
Comment (keichwa) [reply modules/administration/pages/auth-methods-sso.adoc:197]: ok, fine.

File: modules/administration/pages/auth-methods-sso.adoc (hunk 9 above)
[the remark this answers is not in the document]
Comment (keichwa) [reply modules/administration/pages/auth-methods-sso.adoc:117]: the UI here eats HTML-like markup ;)  ```

**Comment by Loquacity** [comment]
@srbarrios waiting on your review so we can merge this

**Comment by srbarrios** [comment]
> @srbarrios waiting on your review so we can merge this

I will review today, sorry for the delay.

File: modules/administration/pages/auth-methods-sso.adoc (hunk 5 above)
Comment (srbarrios) [reply modules/administration/pages/auth-methods-sso.adoc:202]: Yes, you are right @keichwa, my bad, so the point there is that when I did the integration with Keycloak I already had setup a user in SUMA, so on my understanding, to not confuse the customer, we use the same password here.
But in fact, you can use a different password, the only problem is that the customer must know that needs to use that new password. So.. the passwords added on SUMA are not anymore usable.

File: modules/administration/pages/auth-methods-sso.adoc [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+spacewalk-service restart
+----
+
+When you visit the {productname} URL, you will be redirected to the IdP for SSO where you will be requested to authenticate.
+Upon successful authentication, you will be redirected to the {productname} {webui}, logged in as the authenticated user.
+If you encounter problems with logging in using SSO, check the {productname} logs for more information.
+
+
+
+== Example SSO Implementation
```
Comment (paususe) [inline modules/administration/pages/auth-methods-sso.adoc:98]: I would split the Keycloak part to a subpage. In the not-so-distant future, we will also document other SAML providers. One subpage per SAML provider seems more sensible that a long page with lots of information.

File: modules/administration/pages/auth-methods-sso.adoc [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+* Assertion consumer service (or ACS): an endpoint to accept SAML messages to establish a session into the Service Provider.
+The endpoint for ACS in {productname} is: https://example.com/rhn/manager/sso/acs
+* Single logout service (or SLS): an endpoint to initiate a logout request from the IdP.
+The endpoint for SLS in {productname} is: https://example.com/rhn/manager/sso/sls
+* Metadata: an endpoint to retrieve {productname} metadata for SAML.
+The endpoint for metadata in {productname} is: https://example.com/rhn/manager/sso/metadata
+
+
+[IMPORTANT]
```
Comment (paususe) [inline modules/administration/pages/auth-methods-sso.adoc:44]: IMHO we should move the warning above the prerequisites

File: modules/administration/pages/auth-methods-sso.adoc (hunk 5 above)
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:202]: Probably simpler in this case to just say to use the same here. If they're using LDAP it's a different procedure anyway.

File: modules/administration/pages/auth-methods-sso.adoc [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+* Assertion signatures.
+* Message signatures with AuthNRequest, LogoutRequest, and LogoutResponses.
+* Enable an Assertion consumer service endpoint.
+* Enable a single logout service endpoint.
+* Publish the SP metadata (which can be signed).
+
+SSO in {productname} does not support:
+
+* Product choosing and implementation for the identity service provider (IdP).
+* SAML support for other products (check with the respective product documentation).
```
[the remark this answers is not in the document]
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:25]: That's old copy from when we first added SSO. Should we take it out?

File: modules/administration/pages/auth-methods-sso.adoc (hunk 10 above)
Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:98]: Yep, I already split it into PAM and SSO, happy to split the example out again.

**Comment by Loquacity** [comment]
@srbarrios & @paususe can you please confirm that this is OK to merge?

File: modules/administration/pages/auth-methods-sso.adoc [hunk 13]
```diff
@@ -0,0 +1,97 @@
+[[auth-methods]]
+= Authentication With Single Sign-On (SSO)
+
+{productname} supports single sign-on (SSO) by implementing the Security Assertion Markup Language (SAML){nbsp}2 protocol.
```
Comment (srbarrios) [inline modules/administration/pages/auth-methods-sso.adoc:4]: I discovered yesterday that SSO only works for the WebUI, but not for console commands.
We should add a big warning about it.

Comment (Loquacity) [reply modules/administration/pages/auth-methods-sso.adoc:4]: I believe there's a separate issue about that.

**Comment by Loquacity** [comment]
> Apart from the separate issue, I think this PR looks good.

OK, I'll merge this one and do the admon in a new PR.

**Pull request merged**